### PR TITLE
autoflex: json interface

### PR DESCRIPTION
**Pull request by johnsonaj**
### Description

Auto flex understands AWS interface types that convert JSON

`document.Interface` contains the following:
- MarshalSmithyDocument()
- UnmarshalSmithyDocument()

### Relations

Relates #36508

### Output from Acceptance Testing

```console
% make testacc TESTARGS='-run=TestAccOpenSearchServerlessLifecyclePolicy_\|TestAccOpenSearchServerlessLifecyclePolicyDataSource_\|TestAccOpenSearchServerlessAccessPolicy_\|TestAccOpenSearchServerlessAccessPolicyDataSource_' PKG=opensearchserverless

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.22.2 test ./internal/service/opensearchserverless/... -v -count 1 -parallel 20  -run=TestAccOpenSearchServerlessLifecyclePolicy_\|TestAccOpenSearchServerlessLifecyclePolicyDataSource_\|TestAccOpenSearchServerlessAccessPolicy_\|TestAccOpenSearchServerlessAccessPolicyDataSource_ -timeout 360m
--- PASS: TestAccOpenSearchServerlessLifecyclePolicy_disappears (14.07s)
--- PASS: TestAccOpenSearchServerlessAccessPolicyDataSource_basic (14.14s)
--- PASS: TestAccOpenSearchServerlessLifecyclePolicyDataSource_basic (14.16s)
--- PASS: TestAccOpenSearchServerlessAccessPolicy_disappears (14.66s)
--- PASS: TestAccOpenSearchServerlessLifecyclePolicy_basic (15.78s)
--- PASS: TestAccOpenSearchServerlessAccessPolicy_basic (16.23s)
--- PASS: TestAccOpenSearchServerlessLifecyclePolicy_update (20.04s)
--- PASS: TestAccOpenSearchServerlessAccessPolicy_update (20.51s)
PASS
ok  	[github.com/hashicorp/terraform-provider-aws/internal/service/opensearchserverless](http://github.com/hashicorp/terraform-provider-aws/internal/service/opensearchserverless)	25.936s
```

### CODE REVIEW

**Comment by johnsonaj** [comment]
```console
% go test -count=1 ./internal/framework/flex/... -run=TestFlatten -v

--- PASS: TestFlattenFrameworkStringListLegacy (0.00s)
    --- PASS: TestFlattenFrameworkStringListLegacy/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringListLegacy/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringListLegacy/nil_array (0.00s)
--- PASS: TestFlattenFrameworkStringValueMapLegacy (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMapLegacy/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMapLegacy/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMapLegacy/nil_map (0.00s)
--- PASS: TestFlattenFrameworkStringList (0.00s)
    --- PASS: TestFlattenFrameworkStringList/nil_array (0.00s)
    --- PASS: TestFlattenFrameworkStringList/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringList/two_elements (0.00s)
--- PASS: TestFlattenFrameworkStringValueList (0.00s)
    --- PASS: TestFlattenFrameworkStringValueList/nil_array (0.00s)
    --- PASS: TestFlattenFrameworkStringValueList/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueList/two_elements (0.00s)
--- PASS: TestFlattenFrameworkStringValueMap (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMap/nil_map (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMap/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMap/two_elements (0.00s)
--- PASS: TestFlattenSimpleNestedBlockWithStringEnum (0.00s)
    --- PASS: TestFlattenSimpleNestedBlockWithStringEnum/single_nested_empty_value (0.00s)
    --- PASS: TestFlattenSimpleNestedBlockWithStringEnum/single_nested_valid_value (0.00s)
--- PASS: TestFlattenFrameworkStringValueListLegacy (0.00s)
    --- PASS: TestFlattenFrameworkStringValueListLegacy/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueListLegacy/nil_array (0.00s)
    --- PASS: TestFlattenFrameworkStringValueListLegacy/zero_elements (0.00s)
--- PASS: TestFlattenFrameworkStringValueSetLegacy (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSetLegacy/nil_array (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSetLegacy/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSetLegacy/zero_elements (0.00s)
--- PASS: TestFlattenFrameworkStringValueSet (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSet/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSet/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSet/nil_array (0.00s)
--- PASS: TestFlattenComplexSingleNestedBlock (0.00s)
    --- PASS: TestFlattenComplexSingleNestedBlock/single_nested_block_pointer (0.00s)
--- PASS: TestFlattenFrameworkStringMap (0.00s)
    --- PASS: TestFlattenFrameworkStringMap/nil_map (0.00s)
    --- PASS: TestFlattenFrameworkStringMap/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringMap/two_elements (0.00s)
--- PASS: TestFlattenComplexNestedBlockWithStringEnum (0.00s)
    --- PASS: TestFlattenComplexNestedBlockWithStringEnum/single_nested_valid_value (0.00s)
    --- PASS: TestFlattenComplexNestedBlockWithStringEnum/single_nested_zero_value (0.00s)
    --- PASS: TestFlattenComplexNestedBlockWithStringEnum/single_nested_empty_value (0.00s)
--- PASS: TestFlattenSimpleSingleNestedBlock (0.00s)
    --- PASS: TestFlattenSimpleSingleNestedBlock/single_nested_block_value (0.00s)
    --- PASS: TestFlattenSimpleSingleNestedBlock/single_nested_block_pointer (0.00s)
    --- PASS: TestFlattenSimpleSingleNestedBlock/single_nested_block_nil (0.00s)
--- PASS: TestFlattenGeneric (0.00s)
    --- PASS: TestFlattenGeneric/map_block_key_ptr_source (0.00s)
    --- PASS: TestFlattenGeneric/map_block_key_ptr_both (0.00s)
    --- PASS: TestFlattenGeneric/map_string (0.00s)
    --- PASS: TestFlattenGeneric/map_block_enum_key (0.00s)
    --- PASS: TestFlattenGeneric/nested_string_map (0.00s)
    --- PASS: TestFlattenGeneric/non-empty_[]*struct_and_non-empty_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/non-empty_[]*struct_and_non-empty_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/empty_[]*struct_and_empty_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/nil_[]*struct_and_null_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/map_block_key_set (0.00s)
    --- PASS: TestFlattenGeneric/nil_[]*struct_and_null_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/empty_[]*struct_and_empty_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/nil_*struct_Source_and_single_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/nil_[]struct_and_null_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/empty_[]struct_and_empty_struct_Target (0.00s)
    --- PASS: TestFlattenGeneric/empty_[]struct_and_empty_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/nil_[]struct_and_null_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/non-empty_[]struct_and_non-empty_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/*struct_Source_and_single_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/*struct_Source_and_single_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/non-empty_[]struct_and_non-empty_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/complex_Source_and_complex_Target (0.00s)
    --- PASS: TestFlattenGeneric/map_block_key_list (0.00s)
--- PASS: TestFlatten (0.00s)
    --- PASS: TestFlatten/nil_Source_and_Target (0.00s)
    --- PASS: TestFlatten/non-struct_Target (0.00s)
    --- PASS: TestFlatten/non-pointer_Target (0.00s)
    --- PASS: TestFlatten/non-struct_Source (0.00s)
    --- PASS: TestFlatten/single_*string_Source_and_single_string_Target (0.00s)
    --- PASS: TestFlatten/empty_struct_pointer_Source_and_Target (0.00s)
    --- PASS: TestFlatten/single_empty_string_Source_and_single_string_Target (0.00s)
    --- PASS: TestFlatten/empty_struct_Source_and_Target (0.00s)
    --- PASS: TestFlatten/slice/map_of_primtive_types_Source_and_List/Set/Map_of_primtive_types_Target (0.00s)
    --- PASS: TestFlatten/zero_value_slice/map_of_primtive_types_Source_and_List/Set/Map_of_primtive_types_Target (0.00s)
    --- PASS: TestFlatten/capitalization_field_names (0.00s)
    --- PASS: TestFlatten/single_string_Source_and_single_int64_Target (0.00s)
    --- PASS: TestFlatten/does_not_implement_attr.Value_Target (0.00s)
    --- PASS: TestFlatten/resource_name_prefix (0.00s)
    --- PASS: TestFlatten/strange_plurality (0.00s)
    --- PASS: TestFlatten/json_interface_Source_string_Target (0.00s)
    --- PASS: TestFlatten/zero_value_slice/map_of_string_type_Source_and_List/Set/Map_of_string_types_Target (0.00s)
    --- PASS: TestFlatten/single_nil_*string_Source_and_single_string_Target (0.00s)
    --- PASS: TestFlatten/single_nil_*string_Source_and_single_ARN_Target (0.00s)
    --- PASS: TestFlatten/zero_value_primtive_types_Source_and_primtive_types_Target (0.00s)
    --- PASS: TestFlatten/primtive_types_Source_and_primtive_types_Target (0.00s)
    --- PASS: TestFlatten/single_*string_Source_and_single_ARN_Target (0.00s)
    --- PASS: TestFlatten/timestamp (0.00s)
    --- PASS: TestFlatten/timestamp_pointer (0.00s)
    --- PASS: TestFlatten/single_string_Source_and_single_string_Target (0.00s)
    --- PASS: TestFlatten/single_string_Source_and_single_ARN_Target (0.00s)
    --- PASS: TestFlatten/plural_ordinary_field_names (0.00s)
    --- PASS: TestFlatten/single_string_struct_pointer_Source_and_empty_Target (0.00s)
    --- PASS: TestFlatten/timestamp_empty (0.00s)
    --- PASS: TestFlatten/slice/map_of_string_types_Source_and_List/Set/Map_of_string_types_Target (0.00s)
    --- PASS: TestFlatten/timestamp_nil (0.00s)
    --- PASS: TestFlatten/plural_field_names (0.00s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/framework/flex	0.451s
```

**Comment by ewbankkit** [comment]
There also some Smithy document stuff in `internal/json/smithy.go` -- not sure if there's any overlap.

File: internal/framework/flex/auto_flatten.go [hunk 1]
```diff
@@ -92,6 +92,7 @@ func (flattener autoFlattener) convert(ctx context.Context, vFrom, vTo reflect.V
 
 	case reflect.Interface:
 		// Smithy union type handling not yet implemented. Silently skip.
```
Comment (jar-b) [inline internal/framework/flex/auto_flatten.go:94]:
```suggestion
```

File: internal/service/opensearchserverless/lifecycle_policy.go [hunk 2]
```diff
@@ -128,7 +127,11 @@ func (r *resourceLifecyclePolicy) Create(ctx context.Context, req resource.Creat
 	}
 
 	state := plan
-	resp.Diagnostics.Append(state.refreshFromOutput(ctx, out.LifecyclePolicyDetail)...)
+
+	resp.Diagnostics.Append(flex.Flatten(ctx, out.LifecyclePolicyDetail, &state)...)
+
+	state.ID = flex.StringToFramework(ctx, out.LifecyclePolicyDetail.Name)
+	// resp.Diagnostics.Append(state.refreshFromOutput(ctx, out.LifecyclePolicyDetail)...)
```
Comment (jar-b) [inline internal/service/opensearchserverless/lifecycle_policy.go:134]:
```suggestion
```

**Review by YakDriver — APPROVED** [review]
LGTM

File: internal/framework/flex/auto_flatten.go [hunk 3]
```diff
@@ -92,6 +92,7 @@ func (flattener autoFlattener) convert(ctx context.Context, vFrom, vTo reflect.V
 
 	case reflect.Interface:
```
Comment (jar-b) [inline internal/framework/flex/auto_flatten.go:93]: Is there an `auto_expand` equivalent to this case?

**Review by jar-b — APPROVED** [review]
LGTM 🎉

**Comment by johnsonaj** [comment]
> There also some Smithy document stuff in `internal/json/smithy.go` -- not sure if there's any overlap.

@ewbankkit thanks. I was looking for a good location to place the interface type. Could possibly move it there. Also need to think about how `expand` would work

File: internal/framework/flex/auto_flatten.go (hunk 3 above)
Comment (johnsonaj) [reply internal/framework/flex/auto_flatten.go:93]: @jar-b not yet. `Expand` needs a package specific function to work correctly. Still thinking through how to make that work as easily as possible

Comment (jar-b) [reply internal/framework/flex/auto_flatten.go:93]: Makes sense. This is not a blocker (should've indicated as much in the original comment), appreciate the context!

**Pull request merged**